File: web/src/services/test/TestRunService.js
import { request, METHOD } from "@/utils/request";

export function findTestRuns(
  projectId,
  statusId,
  keyword,
  filter,
  pageable,
  sort
) {
  return request({
    url: "/test/run/list",
    method: METHOD.POST,
    params: {
      projectId,
      statusId,
      keyword,
      ...pageable,
      sort,
    },
    data: filter,
  });
}
export function findTestRunsByIds(projectId, testRunIds) {
  return request({
    url: "/test/run/listByIds",
    method: METHOD.GET,
    params: { projectId, testRunIds },
  });
}
export function findOneTestRun(id) {
  return request({
    url: "/test/run/" + id,
    method: METHOD.GET,
  });
}
export function createTestRun(parameter) {
  return request({
    url: "/test/run",
    method: METHOD.POST,
    data: parameter,
  });
}

export function updateTestRun(parameter) {
  return request({
    url: "/test/run",
    method: METHOD.PUT,
    data: parameter,
  });
}
export function deleteTestRun(id) {
  return request({
    url: "/test/run/" + id,
    method: METHOD.DELETE,
  });
}

export function findTestCaseIds(id) {
  return request({
    url: "/test/run/testCaseIds/" + id,
    method: METHOD.GET,
  });
}

export function associateTestCases(testRunId, testCaseIds) {
  return request({
    url: `/test/run/${testRunId}/associate/testCases?`,
    method: METHOD.PUT,
    data: testCaseIds,
  });
}
export function removeTestResults(testRunId, testResultIds) {
  return request({
    url: `/test/run/${testRunId}/remove/testResults?`,
    method: METHOD.PUT,
    data: testResultIds,
  });
}

export function copyTestResults(testRunId, testResultIds) {
  return request({
    url: `/test/run/${testRunId}/copy/testResults?`,
    method: METHOD.PUT,
    data: testResultIds,
  });
}

export function findTestResults(
  projectId,
  testRunId,
  filter,
  keyword,
  pageable,
  sort
) {
  return request({
    url: "/test/run/testResults",
    method: METHOD.POST,
    params: {
      projectId: projectId,
      testRunId: testRunId,
      keyword: keyword,
      ...pageable,
      sort,
    },
    data: filter,
  });
}

export function changeExecutor(testRunId, testResultIds, executorId) {
  return request({
    url: `/test/run/${testRunId}/changeExecutor`,
    method: METHOD.PUT,
    params: {
      executorId,
    },
    data: testResultIds,
  });
}

export function batchChangeResult(
  testRunId,
  testResultIds,
  resultId,
  resultDesc
) {
  return request({
    url: `/test/run/${testRunId}/changeResult`,
    method: METHOD.PUT,
    params: {
      resultId,
      resultDesc,
    },
    data: testResultIds,
  });
}

export function saveResult(
  testRunId,
  testResultId,
  resultId,
  resultDesc,
  testStepResults
) {
  return request({
    url: `/test/run/${testRunId}/saveResult`,
    method: METHOD.PUT,
    params: {
      testResultId,
      resultId,
      resultDesc,
    },
    data: testStepResults,
  });
}

export function linkTestResultWithDownstreamTrackerItems(
  testResultId,
  trackerItemIds
) {
  return request({
    url: `/test/run/trackerTestResultLink/link`,
    method: METHOD.POST,
    params: {
      testResultId,
    },
    data: trackerItemIds,
  });
}

export function unlinkTestResultWithDownstreamTrackerItems(
  testResultId,
  trackerItemIds
) {
  return request({
    url: `/test/run/trackerTestResultLink/unlink`,
    method: METHOD.POST,
    params: {
      testResultId,
      trackerItemIds,
    },
  });
}
export function findDownstreamTrackerItemsByTestResultId(testResultId) {
  return request({
    url: `/test/run/${testResultId}/downstreamTrackerItems`,
    method: METHOD.GET,
  });
}
export function findUpstreamTrackerTestResultsByTrackerItemId(itemId) {
  return request({
    url: `/test/run/${itemId}/upstreamTestResults`,
    method: METHOD.GET,
  });
}

export function findDownstreamTestResultsByTrackerItemId(itemId) {
  return request({
    url: `/test/run/${itemId}/downstreamTestResults`,
    method: METHOD.GET,
  });
}

export function findTestResultsByTrackerItemId(itemId) {
  return request({
    url: `/test/run/${itemId}/testResults`,
    method: METHOD.GET,
  });
}

File: server/helm-test/src/main/java/com/harbortek/helm/test/service/TestRunService.java
/*
 * Copyright [2025] [Harbortek Corp.]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.harbortek.helm.test.service;

import com.harbortek.helm.test.vo.TestResultVo;
import com.harbortek.helm.test.vo.TestRunVo;
import com.harbortek.helm.tracker.vo.items.TrackerItemVo;
import com.harbortek.helm.tracker.vo.tracker.test.TestStepResult;
import com.harbortek.helm.tracker.vo.view.ObjectFilter;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface TestRunService {
    Page<TestRunVo> findTestRuns(Long projectId, Long statusId, String keyword, ObjectFilter filter, Pageable pageable);

    List<TestRunVo> findTestRunByIds(Long projectId,List<Long> testRunIds);

    TestRunVo findOneTestRun(Long id);

    TestRunVo createTestRun(TestRunVo testRunVo);

    TestRunVo updateTestRun(TestRunVo testRunVo);

    void deleteTestRun(Long id);

    List<Long> findTestCaseIds(Long runId);

    void associateTestCases(Long testRunId, List<Long> trackerItems);

    void removeTestResults(Long testRunId, List<Long> resultIds);

    Page<TestResultVo> findTestResults(Long projectId, Long testRunId, ObjectFilter filter, String keyword,
                                       Pageable pageable);

    void batchChangeExecutor(Long testRunId, List<Long> testResultIds, Long executorId);

    void changeResult(Long testRunId, List<Long> testResultIds, Long resultId, String resultDesc);

    TestResultVo saveResult(Long testRunId, Long testResultId, Long resultId, String resultDesc,
                            List<TestStepResult> testStepResults);

    void unlinkTestResultWithDownstreamTrackerItems(Long testResultId);

    void linkTestResultWithDownstreamTrackerItems(Long testResultId, List<Long> trackerItemIds);

    List<TrackerItemVo> findDownstreamTrackerItemsByTestResultId(Long testResultId);

    List<TestResultVo> findUpstreamTestResultsByTrackerItemId(Long itemId);

    List<TestResultVo> findDownstreamTestResultsByTrackerItemId(Long trackerItemId);

    List<TestResultVo> findTestResultsByTrackerItemId(Long trackerItemId);

    void copyTestResults(Long testRunId, List<Long> testResultIds);


}
